core/store/indexstore/indexer.go

```diff
@@ -86,153 +81,86 @@ func (i *Indexer) StoredSection() uint32 {
 	return 0
 }
 
-func (i *Indexer) IsProcessing() bool {
-	return atomic.LoadUint32(&i.processing) == 1
+type LedgerStore interface {
+	GetBlockHash(height uint32) common2.Uint256
+	GetBloomData(height uint32) (types2.Bloom, error)
+	ClearBloomCache(begins, end uint32)
 }
 
-func (i *Indexer) ProcessSection(k store.LedgerStore, interval uint32) error {
-	if atomic.SwapUint32(&i.processing, 1) == 1 {
-		return fmt.Errorf("matcher is already running")
-	}
-	defer atomic.StoreUint32(&i.processing, 0)
-	knownSection := interval / BloomBitsBlocks
+func (i *Indexer) ProcessSection(k LedgerStore, blockHeight uint32) error {
+
+	knownSection := (blockHeight - i.start) / BloomBitsBlocks
 	for i.storedSections < knownSection {
 		section := i.storedSections
-		var lastHead common.Hash
-		if section > 0 {
-			lastHead = i.sectionHead(section - 1)
-		}
 
 		var err error
-
 		// Reset and partial processing
 		i.backend.Reset(section)
 
-		begin := section*BloomBitsBlocks + config.GetAddFilterHeight()
-		end := (section+1)*BloomBitsBlocks + config.GetAddFilterHeight()
+		begin := section*BloomBitsBlocks + i.start
+		end := (section+1)*BloomBitsBlocks + i.start
 
 		for number := begin; number < end; number++ {
-			var (
-				bloom ethtypes.Bloom
-				hash  common.Hash
-			)
-			i.updateBlock()
-			if number == config.GetAddFilterHeight() {
-				bloom = ethtypes.Bloom{}
-				hash = common.Hash{}
-			} else {
-				blockHash := k.GetBlockHash(number)
-				hash = common.Hash(blockHash)
-				if hash == (common.Hash{}) {
-					return fmt.Errorf("canonical block %d unknown", number)
-				}
-
-				bloom, err = k.GetBloomData(number)
-				if err != nil {
-					return fmt.Errorf("get bloom data height: %d %s", number, err.Error())
-				}
+			blockHash := k.GetBlockHash(number)
+			hash := common.Hash(blockHash)
+			if hash == (common.Hash{}) {
+				return fmt.Errorf("canonical block %d unknown", number)
+			}
+
+			bloom, err := k.GetBloomData(number)
+			if err != nil {
+				return fmt.Errorf("get bloom data height: %d %s", number, err.Error())
 			}
-			i.backend.Process(hash, number, bloom)
-			lastHead = hash
+			i.backend.Process(hash, number, bloom, i.start)
```

Comment:
why not call this when persist each block so the bloom cache code can be cleaned.